src/pybind/mgr/dashboard/module.py

```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
-                logger.debug(osd_map)
-                result = {}
+                result = []
                 for osd in osd_map['osds']:
                     osd_id = "{0}".format(osd['osd'])
                     counter_data = global_instance().get_counter('osd', osd_id, path)
-                    result[osd['osd']] = counter_data[path]
+                    result.append(counter_data[path])
+
+                return result
+
+            #Method to find the average of consolidated data
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_data(self, path):
+                data = self.get_counter_allosd(path)
+                result = []
+                osd_count = len(data)
+
+                for i in range(0,len(data[0])):
+
+                      epoch = data[0][i][0]
+                      values = []
+
+                      for osd_id in range(0,osd_count):
+                            values.append(data[osd_id][i][1])
+
+                      result.append(
+                            [epoch, values])
 
                 return result
 
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_avg(self, path):
+                data = self.get_counter_data(path)
+                result = []
+
+                for i in data:
+                    avg = sum(i[1])/len(i[1])
+                    result.append([i[0], avg])
+
+                return result
+
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_stdev(self, path):
+                data = self.get_counter_data(path)
+                result = []
+
+                for i in data:
+                    values = i[1]
+                    avg = sum(values)/len(values)
+                    diff = [(x-avg)**2 for x in values]
+                    var = sum(diff)/len(values)
+                    stdev = math.sqrt(var)
+                    result.append([i[0], stdev])
+
+                return result
+
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_max(self, path):
+                data = self.get_counter_data(path)
+                result = []
+
+                for i in data:
+                    result.append([i[0], max(i[1])])
+
+                return result
+
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_min(self, path):
+                data = self.get_counter_data(path)
+                result = []
+
+                for i in data:
+                    result.append([i[0], min(i[1])])
+
+                return result
+
+            @cherrypy.expose
+            def perf_graph(self, path):
+                template = env.get_template("perf_graphs.html")
+                toplevel_data = self._toplevel_data()
+
+                content_data = {
+                    "avg" : self.get_counter_avg(path),
```

Comment:
no need to `get_counter_data()` repeatedly, please store it locally. and get the avg and stddev in a single run.

Reply:
I admit you are right. I realised the same while I was working on the graphs, so I did replace the entire set of separate methods into a single method get_counter_stats() which stores the data of get_counter_allosd() one time, and then processes it to return the needed values of average, stdevs etc.